pkg/networking/logging_test.go

```diff
@@ -0,0 +1,285 @@
+			levelToLogAt:   zerolog.WarnLevel,
+			expectedResult: false,
+		},
+		{
+			name:           "CurrentLevelEqualToThreshold",
+			currentLevel:   zerolog.InfoLevel,
+			levelToLogAt:   zerolog.InfoLevel,
+			expectedResult: false,
+		},
+		{
+			name:           "TraceLevelBelowDebug",
+			currentLevel:   zerolog.TraceLevel,
+			levelToLogAt:   zerolog.DebugLevel,
+			expectedResult: false,
+		},
+		{
+			name:           "PanicLevelAboveError",
+			currentLevel:   zerolog.PanicLevel,
+			levelToLogAt:   zerolog.ErrorLevel,
+			expectedResult: true,
+		},
+		{
+			name:           "FatalLevelAboveError",
+			currentLevel:   zerolog.FatalLevel,
+			levelToLogAt:   zerolog.ErrorLevel,
+			expectedResult: true,
+		},
+		{
+			name:           "DisabledLevelAboveAll",
+			currentLevel:   zerolog.Disabled,
+			levelToLogAt:   zerolog.TraceLevel,
+			expectedResult: true,
+		},
+		{
+			name:           "DisabledLevelAboveAll",
+			currentLevel:   zerolog.NoLevel,
+			levelToLogAt:   zerolog.TraceLevel,
+			expectedResult: true,
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			result := shouldNotLog(tc.currentLevel, tc.levelToLogAt)
+			assert.Equal(t, tc.expectedResult, result)
+		})
+	}
+}
+
+func Test_LogRequest_NoLog(t *testing.T) {
+	buffer := bytes.Buffer{}
+	logger := zerolog.New(&buffer).Level(zerolog.InfoLevel)
+	request := &http.Request{}
+
+	// method under test
+	LogRequest(request, &logger)
+
+	assert.Empty(t, buffer.Bytes())
+}
+
+func Test_LogRequest_happyPath_server_request(t *testing.T) {
+	expectedBody := "hello world"
+	body := io.NopCloser(bytes.NewBufferString(expectedBody))
+
+	logBuffer := bytes.Buffer{}
+	logger := zerolog.New(&logBuffer).Level(zerolog.TraceLevel)
+	request, err := http.NewRequest(http.MethodGet, "http://localhost/", body)
+	assert.NoError(t, err)
+
+	// method under test
+	LogRequest(request, &logger)
+	assert.Contains(t, logBuffer.String(), expectedBody)
+
+	// ensure that the request body can still be read
+	actualBody, err := io.ReadAll(request.Body)
+	assert.NoError(t, err)
+	assert.Equal(t, expectedBody, string(actualBody))
+}
+
+func Test_LogRequest_happyPath_client_request(t *testing.T) {
+	expectedBody := "hello world"
+	body := io.NopCloser(bytes.NewBufferString(expectedBody))
+
+	logBuffer := bytes.Buffer{}
+	logger := zerolog.New(&logBuffer).Level(zerolog.TraceLevel)
+	request, err := http.NewRequest(http.MethodGet, "http://localhost/", nil)
+	request.GetBody = func() (io.ReadCloser, error) {
+		return body, nil
+	}
+	assert.NoError(t, err)
+
+	// method under test
+	LogRequest(request, &logger)
+	assert.Contains(t, logBuffer.String(), expectedBody)
+}
+
+func Test_LogRequest_happyPath_gzipped_request(t *testing.T) {
+	expectedBody := "hello world"
+	gzipBuffer := bytes.NewBuffer([]byte{})
+	gzipWriter := gzip.NewWriter(gzipBuffer)
+	n, err := gzipWriter.Write([]byte(expectedBody))
+	assert.NoError(t, err)
+	assert.Greater(t, n, 0)
+	err = gzipWriter.Close()
```

Comment:
Pretty nitpicky, but should you maybe defer the `Close()` in case something goes wrong in the writes?

Reply:
I can't really defer the close in this case, because I need it to flush the writer content into the buffer before making the request.